### PR TITLE
BiG-CZ: Analyze Climate

### PR DESCRIPTION
**Pull request by rajadain**
## Overview

Adds climate analysis tab to BiG-CZ. See commit messages for details.

Connects #1949 

### Notes

In some cases, `chord_unlock` isn't called immediately when all the tasks in the geoprocessing operation group ends, as [documented here](https://groups.google.com/forum/#!topic/celery-users/ekzAqp1x-js). But this is hard to reproduce, and does not happen consistently. If this happens often enough, I'll make a card for it.

## Testing Instructions

 * Check out this branch, restart Celery `vagrant ssh worker -c 'sudo service celeryd restart'` and `bundle`
 * Go to [:8000/?bigcz](http://localhost:8000/?bigcz)
 * Select an area of interest. Go to the analyze stage, ensure you see the Climate tab and it has reasonable values. Ensure sorting and downloading of data works correctly.
 * Select a well known area of interest. Ensure it gets cached correctly upon re-run.

### CODE REVIEW

**Comment by ajrobbins** [comment]
Consider changing the column titles to "Mean Precipitation" and "Mean Temperature" to align with the original request and make it totally clear that the values are means.

**Comment by rajadain** [comment]
How does this look?

image

**Comment by ajrobbins** [comment]
Typo in temperature - to avoid the wrapping could we actually do "Mean Precip." and "Mean Temp."?

**Comment by mmcfarland** [comment]
I've been unable to get the climate jobs to complete, on large or small AoIs.  In `debugcelery` I can see the tasks are getting executed up to and including `collect_climate`, but the job doesn't appear to be saved and the polling eventually times out.  Even trying to manually `GET` on that task much later doesn't show it ever completing.  Is this the chord unlock scenario you alluded to?

```
[2017-09-12 10:46:37,612: INFO/MainProcess] Received task: apps.core.tasks.save_job_result[ddc842ce-f46c-471e-8fe4-73177d3a32f3]
[2017-09-12 10:46:37,615: INFO/MainProcess] Task apps.geoprocessing_api.tasks.collect_climate[a7c8d3cd-a212-4066-8864-518ee46051d8] succeeded in 0.0148669000046s: {u'survey': {u'displayName': u'Climate', u'name': u'climate', u'categories': [{u'ppt': 84.76590659401633, u'tmean':...
[2017-09-12 10:46:37,619: DEBUG/MainProcess] Task accepted: apps.core.tasks.save_job_result[ddc842ce-f46c-471e-8fe4-73177d3a32f3] pid:3309
[2017-09-12 10:46:37,650: INFO/MainProcess] Task apps.core.tasks.save_job_result[ddc842ce-f46c-471e-8fe4-73177d3a32f3] succeeded in 0.0333007879963s: None
```

**Comment by rajadain** [comment]
> Is this the chord unlock scenario you alluded to?

Yes. I think Celery has a hard time with that many jobs in parallel (this runs 24, even when the values are cached). I'll spend some more time seeing if I can fix it in this PR rather than wait for a new card. I did eventually get it to work with percussive maintenance though.

> could we actually do "Mean Precip." and "Mean Temp."?

Yes, on it.

**Comment by aufdenkampe** [comment]
That looks great, but it's incomplete.

We want to add:
* Ability to select "Related Layer"
* Line plot, very similar to what we presently have under "Hydrology" output for the Watershed Multiyear Model:
screen shot 2017-09-12 at 10 11 40 am

For "Source", let's expand to "PRISM Climate Group" and link the info to here: http://prism.nacse.org (we'll probably eventually change this to point to the Model My Watershed documentation, once it's merged into MMW.)

**Comment by mmcfarland** [comment]
> Yes. I think Celery has a hard time with that many jobs in parallel 

We could also consider reducing the parallelism by bundling the 12 geop calls inside a single task, per category.  It would increase the single job time, but might cut down on considerable overhead.

**Comment by rajadain** [comment]
Added link to PRISM website, and the related layers:



(the colors in the GIF are off, this is more to show behavior). Going to make a follow up card for adding the chart.

**Comment by mmcfarland** [comment]
I'm not sure if this is a worrisome observation or not, but reloading my worker seems to have resolved the unlock situation for me.

**Comment by rajadain** [comment]
> reloading my worker seems to have resolved the unlock situation

I may have done that at some point too. Just added #2249 to upgrade Celery entirely, which hopefully fixes these and other issues.

File: src/mmw/js/src/analyze/templates/climateTable.html [hunk 1]
```diff
@@ -0,0 +1,23 @@
+<table class="table custom-hover" data-toggle="table">
+    <thead>
+    <tr>
+        <th data-visible="false" data-field="monthidx">Month Index</th>
+        <th data-sortable="true" data-field="month" data-sort-name="monthidx">Month</th>
```
Comment (mmcfarland) [inline src/mmw/js/src/analyze/templates/climateTable.html]: Nice job making months sort by calendar order, not alpha.

**Comment by mmcfarland** [comment]
On the related layers, if anything other than "Coverage" tab is active, the layer is added but the layer picker still shows the previous tab contents, with the coverage tab highlighted.  Works correctly if coverage is already selected.

screenshot from 2017-09-12 12 50 42

**Comment by rajadain** [comment]
> layer picker still shows the previous tab contents, with the coverage tab highlighted

That's a feature, not a bug. It indicates that some layer in that tab is active, as we can have layers from multiple tabs active at the same time. See https://github.com/WikiWatershed/model-my-watershed/blame/develop/src/mmw/sass/components/_layerpicker.scss#L83-L85

**Comment by mmcfarland** [comment]
> That's a feature, not a bug.

I think it behaves like a bug.  If we've just asked the app to turn on the layer, we should activate it so that it shows the displayed month, allows for changing the month, can check the legend, etc.  Especially if that's the behavior users encounter when they activate that layer with other methods.

**Comment by rajadain** [comment]
Ah yes, the month slider is missing. Will look into that.

**Comment by rajadain** [comment]
With the most recent commit 64775d5 the correct layer tab is made active when a relevant layer is enabled. This ensures that the opacity and time sliders are enabled appropriately.

**Comment by rajadain** [comment]
Thanks for the thorough review! Charts will be added in #2250. Going to squash and merge this now.

**Pull request merged**